Published as: Frequent clonal relations between metastases and non-index prostate cancer lesions 
Jeroen Kneppers, Oscar Krijgsman, Monique Melis, Jeroen de Jong, Daniel S. Peeper, Elise Bekers, Henk G. van der Poel, Wilbert Zwart, Andries M. Bergman 
Published January 24, 2019 
Citation Information: JCI Insight. 2019;4(2):e124756. https://doi.org/10.1172/jci.insight.124756.


Special thanks to Oscar Krijgsman for providing initial bulk analysis, heatmap scripts and helpful advice.

### Introduction
This is documentation providing a comprehensive outline of produced data and figures in this project.

### Contents

0. Setup
1. QDNAseq processing of LC-WGS (OK code)
2. CGHcall processing (OK)
3. Patient CN heatmaps 
    + Correlation heatmaps
4. Overview CN heatmaps
    + Topbar annotation
5. Zoomed-in CN heatmaps
6. Clonality analysis
    + 6.1 Density plot
    + 6.2 Boxplot
7. CopywriteR for ChIP

### 0. Setup
Original study was conducted in R version 3.4 with packages that function under it. The packages are used in the chunks below, with own scripts being called from "/Additional_Scripts/".

```{r setup, echo = T}
library(CGHcall)
library(QDNAseq)
library(Clonality)
library(heatmap3)
library(Matrix)
library(RColorBrewer)
library(lattice)
library(Clonality)
library(ggplot2)
library(dplyr)
library(ggpubr)

path = "D:/D Projects/201707-02 Primary PCa vs Lymph node metastases/OKData/"

Sample_ann <- read.table(file="CFMPB293_key.txt", sep="\t", header=T, stringsAsFactors = F)
```

### 1. QDNAseq processing of LC-WGS
Requires package QDNAseq, code counts low-coverage whole genome sequencing data reads in arbitrary bins of 30kb to deduce whether there are copy number alternations (gains or losses) for each patient in the 30 patient cohort.

```{r QDNAseq} 
setwd(path)

bins <- getBinAnnotations(binSize=30)
readCounts <- binReadCounts(bins)

readCountsFiltered<- applyFilters(readCounts, residual=T, blacklist=T, chromosomes=NA)
readCountsFiltered <- estimateCorrection(readCountsFiltered)

noisePlot(readCountsFiltered)
copyNumbers<-correctBins(readCountsFiltered) 

copyNumbersNormalized<-normalizeBins(copyNumbers)
copyNumbersSmooth<-smoothOutlierBins(copyNumbersNormalized)

plot(copyNumbersSmooth)

exportBins(copyNumbersSmooth, file="copyNumbers_Prostate_30Kb.txt")
exportBins(copyNumbersSmooth, file="Prostate_30Kb.igv", format="igv")
```


### 2. CHGcall processing
CGHcall reads the QDNAseq data and normalizes, smoothens and segments the QDNAseq data. There is also a segment normalization step which further normalizes the inherently noisy signal before storing it in a CGHcall calls object containing all CNA information.

```{r CGHcall} 
setwd(path)
Data <- read.table(file="copyNumbers_Prostate_30Kb.txt", sep = "\t", header=T)

raw <- make_cghRaw(Data)
prep <- preprocess(raw, maxmiss = 0, nchrom = 22) 
nor <-  normalize(prep,method = "median", smoothOutliers = TRUE)

seg <-  segmentData(nor, method = "DNAcopy",nperm=2000,undo.splits="sdundo",
                    min.width=5,undo.SD=2, clen=25, relSDlong=5)

segnorm <- postsegnormalize(seg,inter=c(-0.4,0.4))
listcalls <- CGHcall(segnorm,nclass=5,robustsig="yes",cellularity=1,ncpus=12)
calls <- ExpandCGHcall(listcalls,segnorm, divide=5, memeff=FALSE)
save(calls, file="Zwart_Prostate.Rdata")
```

### 3. Patient CN heatmaps
The following loop takes the calls from a certain patient through indexing by the sample annotation file.
It then plots copy number gains and losses per chromosome in an overviewplot, which is a script written by OK.

```{r Plotting heatmaps per patient}
setwd(path)
load("Zwart_Prostate.Rdata")  

Index_samples <- match(paste0("X",Sample_ann[,1]), colnames(calls))
calls <- calls[,Index_samples]

## Double check to see if sorted correctly
table(colnames(calls) == paste0("X",Sample_ann[,1]))

#Plotting loop of CNA per patients
for (i in 1:length(unique(Sample_ann$Patient))){
	IndexSelectedSamples <- which(Sample_ann[,4] == i)
	
	pdf(file=paste0("Patient_",i,".pdf"), width=7, height=4)
	source("/Additional_Scripts/overviewPlot_MIN-MAX.R")
	overviewPlot(calls[,IndexSelectedSamples], scaling=0.6)
	
	dev.off()
}
```

#### 3.1 Correlation heatmaps
The following code was used to plot pearson correlations of different tumor foci found in a single patient.

In this case, patient 11 had samples taken from different areas within the same focus(*), but also had many different foci and a lymph node metastasis. The plot in figure 1E shows that the index focus P1 is largely homogeneous in CNAs, while not correlating with the lymph node metastasis (N) nor the other primary loci (P2-P5).

```{r Plotting correlation heatmap of tumor CNAs of a single patient}
PatientSelect = 11 
IndexSelectedSamples <- numeric(0)


for(i in 1:length(PatientSelect)){
  #Firstloop is if multiple patients are given in PatientSelect
  for (j in 1:length(Sample_annotation$Patient)){
    Selection <- which(Sample_annotation[,4] == PatientSelect[i])
    }
  IndexSelectedSamples <- c(IndexSelectedSamples, Selection)
  }

#Pearson correlation
corpat <- cor(segmented(calls[,IndexSelectedSamples]))

#plotting and saving
pdf(file= "heatmap_p11.pdf", width=4*480, height=2.6*480, res=200)
heatmap3(corpat, method = "ward.D2" , breaks = seq(0, 1, length.out = 1025), 
         scale = "none",showColDendro = F, labRow = Sample_ann$Pair_No[IndexSelectedSamples], 
         labCol = Sample_ann$Pair_No[IndexSelectedSamples])

dev.off()
```


### 4. Overview CNA heatmaps
Code used to generate the large overview plot showing genome-wide CNA in the entire 30 patient cohort.
Topbar containing annotation was generated separately and further edited in Adobe illustrator.

```{r Overview CNA heatmap}
#Initialization
Sample_ann <- read.table(file=paste0(path,"CFMPB293_key.txt"), 
                         sep="\t", header=T, stringsAsFactors = F)
Sample_ann_V2 <- read.table(file=paste0(path,"CFMPB293_keyV3.txt"), 
                            sep="\t", header=T, stringsAsFactors = F)

NVcalls <-calls[,which(Sample_ann_V2$Neo =="0")]
Sample_ann_NV <- Sample_ann_V2[!(Sample_ann_V2$Neo== 1),]
table(colnames(NVcalls)==paste0("X",Sample_ann_NV[,1]))

source("/Additional_Scripts/overviewPlot_MIN-MAX_Xchromosome_correct.R")
overviewPlot(NVcalls, scaling =1)

#Alternative matching
#Index_samples<-match(paste0("X",Sample_ann[,1]), colnames(calls))
#calls<-calls[,Index_samples]
#table(colnames(calls)==paste0("X",Sample_ann[,1]))
```

```{r sorting the overviewplot by specific clinical parameters}
#Tumor type sort
overviewPlot(NVcalls[,which(Sample_ann_NV$Tumor=="P")])
overviewPlot(NVcalls[,which(Sample_ann_NV$Tumor=="LN")])
overviewPlot(NVcalls[,which(Sample_ann_NV$Dupli_P== 0)])

#Gleason sort only
NVcalls_sort <- NVcalls[,order(Sample_ann_NV$Gleason, decreasing = T)]
Sample_ann_NV_sort <- Sample_ann_NV[order(Sample_ann_NV$Gleason, decreasing = T),]
table(colnames(NVcalls_sort)==paste0("X",Sample_ann_NV_sort[,1]))

#Tumor type sort only
NVcalls_sort <- NVcalls[,order(Sample_ann_NV$Tumor, decreasing = F)]
Sample_ann_NV_sort <- Sample_ann_NV[order(Sample_ann_NV$Tumor, decreasing = F),]
table(colnames(NVcalls_sort)==paste0("X",Sample_ann_NV_sort[,1]))
Ind <- which(Sample_ann_NV$Neo == 0)

#Combination Gleason & Tumor type
NVcalls_sort <- NVcalls[,order(Sample_ann_NV$Gleason, decreasing = T)]
Sample_ann_NV_sort <- Sample_ann_NV[order(Sample_ann_NV$Gleason, decreasing = T),]
NVcalls_sort <- NVcalls_sort[,order(Sample_ann_NV$Tumor, decreasing = F)]
Sample_ann_NV_sort <- Sample_ann_NV_sort[order(Sample_ann_NV$Tumor, decreasing = F),]

table(colnames(NVcalls_sort)==paste0("X",Sample_ann_NV_sort[,1]))
Ind <- which(Sample_ann_NV$Neo == 0)

overviewPlot(NVcalls_sort[ChrRange, Ind], scaling=1) +
  title(main = paste0("Chromosome ", Sel[1], " ", (Sel[2]/1000), " kb to ", (Sel[3]/1000), " kb"))
```


#### 4.1 Topbar annotation
```{r Topbar visualization} 
#Selection
Bar_ann <- Sample_ann_NV #[which(Sample_ann_NV$Dupli_P == "0"),]
Bar_ann <- Bar_ann[4:8]

for(i in 1:length(Bar_ann$Patient)){
  if (Bar_ann[i,2] == "P"){
    Bar_ann[i,2] <- 0
  }
  else if (Bar_ann[i,2] == "LN"){
    Bar_ann[i,2] <- 1
  }
}

Bar_annM <- matrix(Bar_ann)
steps <- rep(brewer.pal(5, 'RdYlBu'),6)

source(".\\Additional_Scripts\\color_palette.R")
my_cols <- color.palette(steps, c(rep(c(1, 0), 14),1))

par(mar=c(3,1,4,1))
image(Bar_annM, col.regions = my_cols(ncol(Bar_annM)), cuts = ncol(Bar_annM), sub = "", ylab = "", xlab = "", useRaster = F)

#Optional: Order for Gleason or LN
# Bar_ann <- Bar_ann[order(Bar_ann$Gleason, decreasing = T),]
# Bar_ann <- Bar_ann[order(Bar_ann$Tumor, decreasing = T),]
# Bar_ann <- matrix(as.numeric(unlist(Bar_ann)), ncol = ncol(Bar_ann), byrow = F)
# Bar_ann <- t(Bar_ann)
```

### 5. Zoomed-in CN heatmaps
Insets were generated by zooming in at PCa specific loci to visualize CNA in tumors for those loci.

```{r Insets zoomed on specific loci}
#Zoomed AR Initialization, Sel contains genomic coordinates in c(chr,start,end) format
Sel <- c(23, 65000000, 68000000) #Also plotted inset zooms for MYC and TMPRSS2 locations
Ind <- which(Sample_ann_NV$Dupli_P == 0)
ChrRange <- which(chromosomes(NVcalls)== Sel[1] & bpstart(NVcalls)< Sel[3] & bpend(NVcalls) > Sel[2])

source("/Additional_Scripts/overviewPlot_MIN-MAX_Xchromosome_correct.R")
overviewPlot(NVcalls[ChrRange, Ind], scaling=1) +
  title(main = paste0("Chromosome ", Sel[1], " ", (Sel[2]/1000), " kb to ", (Sel[3]/1000), " kb"))

#Sorted ChrRange over random:
#ChrRange_sorted <- rank(-(NVcalls[ChrRange, Ind]))
```

### 6. Clonality analysis
Package by Ostravanaya et al. 2011 Bioinformatics, doi: 10.1093/bioinformatics/btr267

Set WD, load data, sort for Sample_annotation
```{r Loading call object for clonality calculation, echo = FALSE}
setwd(path)
load("Zwart_Prostate.Rdata") 

# Order calls according to patient ID
Sample_ann <- read.table(file="CFMPB293_key.txt", sep="\t", header=T)
Index_samples <- match(paste0("X",Sample_ann[,1]), colnames(calls))
calls <- calls[,Index_samples]
dataCNA <- CNA(copynumber(calls),chrom=chromosomes(calls),maploc=bpstart(calls),sampleid=colnames(calls))

# Average the datapoints, see Clonality Documentation for more information
# New number of datapoints (2150) = bins (86) / averageing parameter (40)
dataAve <- ave.adj.probes(dataCNA,40)
```

Clonality Analysis - Warning - this can take a while dependent on averageing parameter and computing strength
Outputs a histogram of reference vs all results log likelihood relationships (including primary - primary pairs)
```{r Clonality analysis and plotting of log likelihood}
ptlist <- Sample_ann$Patient
results <- clonality.analysis(dataAve, ptlist,  pfreq = NULL, refdata = NULL, nmad = 1,  reference = TRUE, allpairs = TRUE)

histogramPlot(results$LR[,4], results$refLR[,4])
```

#### 6.1 Density Plot
Select only pairs with LN Mets as reference. True checks indicate succesful GCF_ID - Pair_No swapping

```{r Density plots of LLR, echo = FALSE}
#Find index of LNs, initialize for subsetting
for(i in 1:length(Sample_ann$Tumor)){
  Index_LN <- which(Sample_ann$Tumor == "LN")
}

res_df <- results$LR
ref_df <- results$refLR
LNsort <- res_df[FALSE,]

#Subset for Metastases for LR set, then swap GCF_ID for Pair_No in table
for(i in 1:length(Index_LN)){
  LNsort <- rbind(LNsort, subset(res_df, res_df[,2] == paste0("X",Sample_ann$GCF_ID[(Index_LN[i])])))
}

for(i in 1:2){
  Check <- match(LNsort[,i], paste0("X", Sample_ann[,1]))
  LNsort[,i] <- Sample_ann$Pair_No[Check]
}
```

Sorting for cut-off pvalue 0.01
```{r density plot}
#intializing plotting dataframes
ref <- ref_df[,4]
res <- LNsort[,4]

refd <- data.frame(value = ref)
resd <- data.frame(res)

#sorting for cut-off pvalue of 0.01
res <- LNsort
res <- res[order(LNsort[,10]),]

LNsig <- res[which(res[,10] <= 0.01),]
LNns <- res[which(res[,10] >= 0.01),] 

primd<- data.frame(value = LNsig[,4])
smallp <- data.frame(value = LNns[,4])

a <- ggplot() + geom_density(data = log(refd), aes(log(refd)), fill = "grey") + 
  geom_density(data = log(smallp), aes(log(smallp)), color = "navy", fill = "navy", alpha = 0.5) + 
  geom_density(data = log(primd), aes(log(primd)), color = "firebrick", fill = "firebrick", alpha = 0.5) +
  geom_vline(xintercept =  log(4.5e+01)) + labs(x = "log Likelihood Ratio", y = "Density") +
  scale_x_continuous(expand = c(0, 0), limits = c(-10, 34)) + scale_y_continuous(expand= c(0,0), limits = c(0, 0.21)) +
  theme_bw() + theme(axis.text=element_text(size=rel(1.)), axis.title=element_text(size=rel(1.5)))
a
```

#### 6.2 Boxplot
```{r boxplot, requires ggpubr}
refbox <- data.frame(group = "Reference", value = log(refd))
primbox <- data.frame(group = "Dominant P - LN", value = log(primd))
smallbox <- data.frame(group = "Other P - LN", value = log(smallp))
Comp <- rbind(refbox, primbox, smallbox)
color <- c("grey", "firebrick", "navy")

b <- ggplot(Comp, aes(x=group, y= value, fill = group)) + geom_boxplot(alpha = 0.5) + 
  scale_fill_manual(values = color) +
  stat_compare_means(comparisons = list(c("Dominant P - LN", "Other P - LN"))) +
  theme_bw() 
b
```
